pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGateFunctors.hpp

```diff
@@ -671,34 +722,52 @@ void applyCZ(Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
              const std::vector<std::size_t> &wires,
              [[maybe_unused]] const bool inverse = false,
              [[maybe_unused]] const std::vector<PrecisionT> &params = {}) {
-    applyNC2Functor(
-        ExecutionSpace{}, arr_, num_qubits, wires,
-        KOKKOS_LAMBDA(Kokkos::View<Kokkos::complex<PrecisionT> *> arr,
-                      const std::size_t i00, const std::size_t i01,
-                      const std::size_t i10, const std::size_t i11) {
-            [[maybe_unused]] const auto i00_ = i00;
-            [[maybe_unused]] const auto i01_ = i01;
-            [[maybe_unused]] const auto i10_ = i10;
-            arr(i11) *= -1;
-        });
+    auto core_function = KOKKOS_LAMBDA(
+        Kokkos::View<Kokkos::complex<PrecisionT> *> arr, const std::size_t i00,
+        const std::size_t i01, const std::size_t i10, const std::size_t i11) {
+        [[maybe_unused]] const auto i00_ = i00;
+        [[maybe_unused]] const auto i01_ = i01;
+        [[maybe_unused]] const auto i10_ = i10;
+        arr(i11) *= -1;
+    };
+    applyNC2Functor<PrecisionT, decltype(core_function), false>(
+        ExecutionSpace{}, arr_, num_qubits, wires, core_function);
+}
+
+template <class ExecutionSpace, class PrecisionT>
+void applyNCSWAP(Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
+                 const std::size_t num_qubits,
+                 const std::vector<std::size_t> &controlled_wires,
+                 const std::vector<bool> &controlled_values,
+                 const std::vector<std::size_t> &wires,
+                 [[maybe_unused]] const bool inverse = false,
+                 [[maybe_unused]] const std::vector<PrecisionT> &params = {}) {
+    auto core_function = KOKKOS_LAMBDA(
+        Kokkos::View<Kokkos::complex<PrecisionT> *> arr, const std::size_t i00,
+        const std::size_t i01, const std::size_t i10, const std::size_t i11) {
+        [[maybe_unused]] const auto i00_ = i00;
+        [[maybe_unused]] const auto i11_ = i11;
+
+        kokkos_swap(arr(i10), arr(i01));
+    };
+    if (controlled_wires.empty()) {
```

Comment:
Simulations mostly happen in the else branch right?

Reply:
In general in algorithms might use just SWAP more than controlled-SWAP? I'm not sure

Reply:
`SWAP` calls this API with `controlled_wires = {}`